### PR TITLE
DAOS-2354 Placement: Reintegration Placement API and Test

### PR DESCRIPTION
**Pull request by PetFet**
- Implementation of the reintegration API and tests for both ring map and jump map. 
- Added Scons rules for compiling the tests into separate binaries.

### CODE REVIEW

File: src/include/daos/pool_map.h [hunk 1]
```diff
@@ -245,7 +245,8 @@ static inline bool
 pool_target_unavail(struct pool_target *tgt)
```
Comment (NiuYawei) [inline src/include/daos/pool_map.h]: add a 'for_reint' parameter, and regard 'UP' as available when 'for_reint' is true.

File: src/placement/jump_map.c [hunk 2]
```diff
@@ -832,10 +842,96 @@ jump_map_obj_find_rebuild(struct pl_map *map, struct daos_obj_md *md,
...
+	reint_layout->ol_grp_nr = jop.jmop_grp_nr;
+	reint_layout->ol_grp_size = jop.jmop_grp_size;
+	layout->ol_grp_nr = jop.jmop_grp_nr;
+	layout->ol_grp_size = jop.jmop_grp_size;
+
+
+	D_INIT_LIST_HEAD(&remap_list);
+	D_INIT_LIST_HEAD(&reint_list);
+
+	/* Find targets affected directly by reintegration. */
```
Comment (NiuYawei) [inline src/placement/jump_map.c]: The comment is bit confusing, it looks to me it's for getting old layout.

File: src/placement/jump_map.c [hunk 3]
```diff
@@ -832,10 +842,96 @@ jump_map_obj_find_rebuild(struct pl_map *map, struct daos_obj_md *md,
...
+	rc = get_object_layout(jmap, layout, &jop, &remap_list, false, md);
+	if (rc)
+		goto out;
+
+	/* Clear list for next placement operation. */
+	remap_list_free_all(&remap_list);
+	D_INIT_LIST_HEAD(&remap_list);
+
+	/* Find targets indirectly affected by reintegration. */
```
Comment (NiuYawei) [inline src/placement/jump_map.c]: I think it's to get new layout (with reintegrating targets), right?

File: src/placement/pl_map_common.c [hunk 4]
```diff
@@ -63,6 +63,50 @@ remap_add_one(d_list_t *remap_list, struct failed_shard *f_new)
...
 
+void
+reint_add_one(d_list_t *remap_list, struct failed_shard *f_new)
+{
+	struct failed_shard     *f_shard;
+	d_list_t                *tmp;
+
+	D_DEBUG(DB_PL, "Reint target : fnew: %u", f_new->fs_shard_idx);
+
+	/* All failed shards are sorted by fseq in ascending order */
```
Comment (NiuYawei) [inline src/placement/pl_map_common.c]: Looks it's not in fseq order, and I don't think it's necessary in fseq order.

File: src/placement/pl_map_common.c [hunk 5]
```diff
@@ -63,6 +63,50 @@ remap_add_one(d_list_t *remap_list, struct failed_shard *f_new)
...
+	d_list_for_each_prev(tmp, remap_list) {
+		f_shard = d_list_entry(tmp, struct failed_shard, fs_list);
+
+		D_DEBUG(DB_PL, "fnew: %u, fshard: %u", f_new->fs_shard_idx,
+			f_shard->fs_shard_idx);
+
+		if (f_new->fs_shard_idx < f_shard->fs_shard_idx)
+			continue;
+		if (f_new->fs_shard_idx == f_shard->fs_shard_idx)
+			return;
```
Comment (NiuYawei) [inline src/placement/pl_map_common.c]: what's the purpose of keeping the list in shard_idx order?

File: src/placement/pl_map_common.c (hunk 4 above)
Comment (PetFet) [reply src/placement/pl_map_common.c]: Correct, this is important as part of rebuild and finding fallback shards but is unnecessary as part of reintegration.

Will remove.

File: src/placement/pl_map_common.c (hunk 5 above)
Comment (PetFet) [reply src/placement/pl_map_common.c]: See above.

File: src/placement/pl_map_common.c [hunk 6]
```diff
@@ -76,21 +95,38 @@ remap_alloc_one(d_list_t *remap_list, unsigned int shard_idx,
...
+   * Allocate a new failed shard then add it into remap list
+   *
+   * \param[in] remap_list        List for the failed shard to be added onto.
+   * \param[in] shard_idx         The shard number of the failed shard.
+   * \paramp[in] tgt              The failed target that will be added to the
+   *                              remap list.
+   */
+int
+reint_alloc_one(d_list_t *remap_list, unsigned int shard_idx,
+		struct pool_target *tgt)
```
Comment (gnailzenh) [inline src/placement/pl_map_common.c]: do we actually need a separate function, or just add a parameter to say it's "reint" or "rebuild"

File: src/placement/ring_map.c [hunk 7]
```diff
@@ -123,8 +123,8 @@ ring_comp_shuff_cmp(struct pool_component *comp_a,
 		    struct pool_component *comp_b,
 		    unsigned int seed, unsigned int prime)
 {
-	uint64_t	key_a = comp_a->co_id;
-	uint64_t	key_b = comp_b->co_id;
+	uint64_t        key_a = comp_a->co_id;
+	uint64_t        key_b = comp_b->co_id;
```
Comment (gnailzenh) [inline src/placement/ring_map.c]: why do you change all tabs to spaces? I really prefer to use tabs and only touch the code which is supposed to be changed

**Comment by PetFet** [comment]
Known rebuild failure https://jira.hpdd.intel.com/browse/DAOS-3931.

**Comment by PetFet** [comment]
> Why there are so many irrelevant changes? It's caused by some merge errors?

There were some merge conflicts while I was trying to track down some bugs.
I can squash everything after the last review if that will help. Sorry for the messy change history.

**Pull request merged**